shell/platform/darwin/macos/framework/Source/FlutterThreadSynchronizer.mm

```diff
@@ -0,0 +1,132 @@
+#import "flutter/shell/platform/darwin/macos/framework/Source/FlutterThreadSynchronizer.h"
+
+#import <QuartzCore/QuartzCore.h>
+#include <mutex>
+
+@interface FlutterThreadSynchronizer () {
+  std::mutex _mutex;
+  BOOL _shuttingDown;
+  CGSize _contentSize;
+  std::vector<dispatch_block_t> _scheduledBlocks;
+
+  BOOL _beginResizeWaiting;
+
+  // Used to block [beginResize:].
+  std::condition_variable _condBlockBeginResize;
+}
+
+@end
+
+namespace {
+/// Single use event (can only be signalled once)
+class Event {
+ public:
+  void Signal() {
+    assert(!signaled_);
+    std::scoped_lock locker(mutex_);
+    signaled_ = true;
+    cv_.notify_one();
+  }
+
+  void Wait() {
+    std::unique_lock<std::mutex> locker(mutex_);
+    while (!signaled_) {
+      cv_.wait(locker);
+    }
+  }
+
+ private:
+  std::condition_variable cv_;
+  std::mutex mutex_;
+  bool signaled_ = false;
+};
+}  // namespace
+
+@implementation FlutterThreadSynchronizer
+
+- (void)drain {
+  assert([NSThread isMainThread]);
+
+  [CATransaction begin];
+  [CATransaction setDisableActions:YES];
+  for (dispatch_block_t block : _scheduledBlocks) {
+    block();
+  }
+  [CATransaction commit];
+  _scheduledBlocks.clear();
+}
+
+- (void)blockUntilFrameAvailable {
+  std::unique_lock<std::mutex> lock(_mutex);
+
+  _beginResizeWaiting = YES;
+
+  while (CGSizeEqualToSize(_contentSize, CGSizeZero) && !_shuttingDown) {
+    _condBlockBeginResize.wait(lock);
+    [self drain];
+  }
+
+  _beginResizeWaiting = NO;
+}
+
+- (void)beginResize:(CGSize)size notify:(nonnull dispatch_block_t)notify {
+  std::unique_lock<std::mutex> lock(_mutex);
+
+  if (CGSizeEqualToSize(_contentSize, CGSizeZero) || _shuttingDown) {
+    // No blocking until framework produces at least one frame
+    notify();
+    return;
+  }
+
+  [CATransaction begin];
+  [CATransaction setDisableActions:YES];
```

Comment:
I think the transaction in beginResize is not necessary. We might end up in multiple transactions from drain during beginResize:, but that's all in same run-loop turn so it shouldn't matter.

I'd leave the one in drain, since I think it is better to do it there than on two completely separate places (updating CALayers and mutating platform views).